### DISCUSSION TITLE
Add support for Inline Snapshots

### DISCUSSION THREAD

**Discussion by michael-wolfenden** [VerifyTests/Verify#1090] [Ideas]
Hi👋,

Recently a library was announced that allows for inline snapshots:
https://www.meziantou.net/inline-snapshot-testing-in-dotnet.htm

I was wondering if you had considered adding support for inline snapshots in Verify.

Cheers

**Comment by SimonCropp** [comment]
I think its  great idea. but i dont know how to make it work for multiple tests.

For example, 

 * lets say you have 2 inline tests in one file. 
 * you run the tests the first time. 
 * two diff popup with proposed changes to the same file
 * you accept the first, the cs file updates
 * you accept the second, the cs file updates again, but it overides the change from the fist diff.
 * you run again and the first test fails a second time, even though you accepted the change

**Reply by isidore** [reply]
I've been using it in the python and java version of approvaltests. 
I come to really appreciate it form approvals that are less than 10 lines of text.

The key to getting it to work smoothly with multiple failures is to have the reporters run serially. 
this can be a bit of a problem if the parellation is coming from the top. 
But is manageable none the less.
You can also make a multi-file reporter. Meaning if 5 things fail from the same file, the reporter batches them up.
The way you do this is basically to append the failure.
InlineReporter.AppendFailure()
Such that if a,b,e,i all fail, the reporter actually generates 4 times (a, ab, abe, abei) and the last one is shown.
https://github.com/approvals/ApprovalTests.Net/blob/master/src/ApprovalTests/Reporters/AllFailingTestsClipboardReporter.cs
does a simular thing.

**Reply by SimonCropp** [reply]
i dont think that is possible in the .net test ecosystem. we are not dealing with one runtime. when targetting multiple runtimes, each group of tests runs under its own process. u can block other processes. nor can u batch up reporting

**Reply by SimonCropp** [reply]
> You can also make a multi-file reporter. Meaning if 5 things fail from the same file, the reporter batches them up.
The way you do this is basically to append the failure.

I dont think this works. each failure includes the proposed new file in its entirety. so they would still blat the previous change on each accepting

**Reply by meziantou** [reply]
In the case of [Meziantou.Framework.InlineSnapshotTesting](https://www.meziantou.net/inline-snapshot-testing-in-dotnet.htm), the behavior is the following:
- When a snapshot is updated, it creates a temp file with the new content and open the diff tool
- If another test is edited, the temp file is updated. Most diff tools reload the file automatically, so if you haven't accepted the first diff yet, you will see both diffs.

There are limitations
- Snapshot must be updated by the same process. The lib stores a list of all edits to be able to update multiple snapshots in the same file. This list is stored in a field, so it is only available in the current process. Also, the temp file is unique per process, so if 2 processes are updated the same file, you may have issues.
- If the content of the file changed between the compilation and the update, the lib will not be able to find where the snapshot is located in the source file
- In complex cases, you may need to run your tests twice to update all snapshots

If you want to see the implementation:
- [FileEditor.cs](https://github.com/meziantou/Meziantou.Framework/blob/main/src/Meziantou.Framework.InlineSnapshotTesting/FileEditor.cs): Edit the cs file content
- [SnapshotUpdateStrategy.cs](https://github.com/meziantou/Meziantou.Framework/blob/main/src/Meziantou.Framework.InlineSnapshotTesting/SnapshotUpdateStrategy.cs): List available update strategies
- How it can find which lines to edit? <https://www.meziantou.net/inline-snapshot-testing-in-dotnet.htm#how-does-it-know-whi>

**Reply by SimonCropp** [reply]
happy to consider a pull request
